// app/src/map.js
/** @jsx React.DOM */
var React = require('react');
var renderMap = require('app/render');
var d3ify = require('app/d3ify');
var d3 = require('d3');
require('d3-tip');
var tooltipView = require('app/tooltip-view');

module.exports = React.createClass({
  getInitialState: function () {
    return {
      updateable: false,
      force: null,
      init: false
    };
  },

  render: function () {
    return  <svg
              id={this.props.id}
              width="100%"
              height="100%"
              preserveAspectRatio="xMidYMid meet"
              viewBox={"0 0 " + this.props.width + " " + this.props.height} >
              <g id="zoom">
                <g id="mapG">
                  <rect
                    className="overlay"
                    width={this.props.width}
                    height={this.props.height} >
                  </rect>
                </g>
              </g>
            </svg>;
  },

  componentDidMount: function () {
    console.log('map rendering', this.props)
    if (Object.keys(this.props.data.nodes).length > 0) {
      var data = this.props.data
       ,  options = {
            selector: this.props.selector,
            width: this.props.width,
            height: this.props.height
          };

      renderMap(d3ify(data), options, this);
    }    
  },

  registerForceListeners: function () {
    this.state.force.on('end', function () {
      this.setState({updateable: true});
    }.bind(this));
  },

  componentDidUpdate: function (prevProps) {
    console.log('map - componentDidUpdate', this.state, this.props)

    //init map
    if (!this.state.init && Object.keys(this.props.data.nodes).length > 0) {
      var data = this.props.data
       ,  options = {
            selector: this.props.selector,
            width: this.props.width,
            height: this.props.height
          };

      console.log('initingg', data, options)
      renderMap(d3ify(data), options, this);
      this.setState({init: true});
    }

    //render injects force onto map component
    if (this.state.force && !this.state.updateable) {
      this.registerForceListeners();
    };

    //once force has ended animation map becomes updateable
    //TODO live update new nodes
    if (this.state.updateable && this.state.init) {
      var data = d3ify(this.props.data);
      //NOTE React or D3 doesn't allow selection of nodes directly through d3.selectAll('.node')
      // for some reason
      var map = d3.select('#mapG') 

      map .selectAll('.node')
          .data(data.nodes, function(d) { return d.id })
          .classed('open', function(d) { return d.tabId });   
    };
  }
})
