// app/src/star.js
/** @jsx React.DOM */
var React = require('react');

module.exports = React.createClass({
  //TODO remove state from this class and place in higher level 'popup' parent component.
  getInitialState: function () {
    return {
      rank: this.props.rank
    }
  },

  render: function () {
    var width = this.props.width + "px";
    var height= this.props.height + "px";
    var viewBox = "0 0 " + this.props.width  + " " + this.props.height;
    var waypointClass = this.state.rank === 1 ? "waypoint-active" : "waypoint";

    return  <a
              onClick={this.handleClick} 
              className={waypointClass} >
              <svg 
                width={width} 
                height={height}
                viewBox={viewBox}
                version="1.1">
                  <polygon 
                    points="8 11.8125 3.29771798 14.2460088 4.19577393 9.09175442 0.39154787 5.44149117 5.64885899 4.68949558 8 0 10.351141 4.68949558 15.6084521 5.44149117 11.8042261 9.09175442 12.702282 14.2460088 ">
                  </polygon>
              </svg>
            </a>;
  },

  handleClick: function () {
    var toggleInteger = this.state.rank === 1 ? -1 : 1;
    this.setState({rank: toggleInteger});
    chrome.runtime.sendMessage({
      action: 'updateNode',
      nodeId: this.props.id,
      props: {rank: toggleInteger}
    });
    return false;
  }
});
