// app/src/trail-title.js
/** @jsx React.DOM */
var React = require('react');

module.exports = React.createClass({
  getInitialState: function () {
    //TODO remove state from this class and place in higher level 'popup' parent component.
    return {
      editable: false,
      tempValue: this.props.title
    };
  },

  render: function () {
    var editable = this.state.editable;

    if (editable) {
      return <input
              type="text"
              autoFocus
              value={this.state.tempValue}
              onChange={this.handleChange}
              onBlur={this.handleBlur}
              onFocus={this.handleFocus}
              onKeyPress={this.handleKeyPress} />;
    } else {
      return <a href="#" onClick={this.handleIconClick}>
                <span>{this.state.tempValue}</span>
                <img
                  onClick={this.handleIconClick}
                  src="/ui/icons/editable-icon.svg" />
              </a>;
    };
  },

  handleFocus: function(evt) {
    evt.target.select();
  },

  handleKeyPress: function(evt) {
    if (evt.key === 'Enter') this.handleBlur();
    if (evt.key === 'Escape') {
      this.setState({tempValue: this.props.title});
      this.setState({editable: false});
      this.handleBlur();
    }
  },

  handleIconClick: function () {
    this.setState({editable: true});
  },

  handleChange: function (evt) {
    this.setState({tempValue: evt.target.value});
  },

  handleBlur: function (evt) {
    this.setState({editable: false});
    if (this.state.tempValue !== this.props.title) {
      chrome.runtime.sendMessage({
        action: 'updateAssignment',
        assignmentId: this.props.id,
        props: {title: this.state.tempValue}
      });
    };
  }
});
