// app/src/trail-title.js
/** @jsx React.DOM */
var React = require('react');

module.exports = React.createClass({
  getInitialState: function () {
    //possible antipattern of props in state
    return {
      editable: false,
      initValue: this.props.title
    };
  },

  render: function () {
    var editable = this.state.editable;

    if (editable) {
      return <div> 
                <input 
                type="text" 
                value={this.state.initValue}
                onChange={this.handleChange}
                onBlur={this.handleBlur} />
                <img src="/ui/icons/editable-icon.svg" />
              </div>
    } else {
      return  <h1>
                <a href="#">
                  {this.state.initValue}
                  <img
                    onClick={this.handleIconClick} 
                    src="/ui/icons/editable-icon.svg" />
                </a>
              </h1>;
    };
  },

  handleIconClick: function () {
    this.setState({editable: true})
  },

  handleChange: function (evt) {
    this.setState({initValue: evt.target.value})
  },

  handleBlur: function (evt) {
    console.log('blur', evt, this.state)
    this.setState({editable: false});
    chrome.runtime.sendMessage({
      action: 'updateAssignment', 
      assignmentId: this.props.id, 
      newTitle: this.state.initValue
    });
  }
});
