// app/ext-node-click.js
var d3 = require('d3');

module.exports = function(node) {
  // If a link is middle clicked or ctrl/cmd+clicked
  if (d3.event.which === 2 || (d3.event.which === 1 && (d3.event.metaKey || d3.event.ctrlKey))) {
    d3.event.preventDefault();
    d3.event.stopPropagation();

    chrome.runtime.sendMessage({
      action: "trackUIEvent",
      eventName: "ui.map.node.click",
      eventData: {
        button: "meta"
      }
    });

    // Tell the runtime to open a new tab
    chrome.runtime.sendMessage({
      action: 'resumeAssignment',
      nodeId: node.id
    });
  } else if (d3.event.which === 1) {
    d3.event.preventDefault();
    d3.event.stopPropagation();

    chrome.runtime.sendMessage({
      action: "trackUIEvent",
      eventName: "ui.map.node.click",
      eventData: {
        button: "left"
      }
    });

    if (node.tabId) {
      chrome.tabs.update(node.tabId, { active: true }, function(tab) {
        chrome.windows.update(tab.windowId, {focused: true});
      });
    } else {
      // Tell the runtime to open a new tab
      chrome.runtime.sendMessage({
        action: 'resumeAssignment',
        nodeId: node.id,
        focus: true
      });
    }
  }
};
